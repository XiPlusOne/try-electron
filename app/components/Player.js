// @flow
import React from 'react';
import soundTrack from '../../resources/bg.flac';

type Props = {
  onMusicEnd: () => void,
  isPlaying: boolean
};

export default class Player extends React.PureComponent<Props> {
  refAudio: $Call<typeof React.createRef>;

  constructor(props: Props) {
    super(props);

    this.refAudio = React.createRef();
  }

  componentDidMount() {
    const audioContext = new AudioContext();
    const { current } = this.refAudio;
    if (current) {
      const track = audioContext.createMediaElementSource(current);
      track.connect(audioContext.destination);
    }
  }

  componentDidUpdate(prevProps: Props) {
    const { isPlaying } = this.props;
    const { isPlaying: wasPlaying } = prevProps;

    if (isPlaying === wasPlaying) {
      return;
    }

    if (isPlaying) this.playMusic();
    else this.pauseMusic();
  }

  playMusic = () => {
    const { current } = this.refAudio;
    if (current) {
      current.play();
    }
  };

  pauseMusic = () => {
    const { current } = this.refAudio;
    if (current) {
      current.pause();
    }
  };

  render() {
    const { onMusicEnd } = this.props;

    return (
      <audio
        src={soundTrack}
        preload="auto"
        ref={this.refAudio}
        onEnded={onMusicEnd}
      >
        <track default kind="captions" label="Hello World!" />
      </audio>
    );
  }
}
