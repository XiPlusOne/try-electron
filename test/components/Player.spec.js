import { spy } from 'sinon';
import React from 'react';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import Player from '../../app/components/Player';

Enzyme.configure({ adapter: new Adapter() });

/* eslint-disable */
class AudioContext {
  createMediaElementSource() {
    return new MediaElementAudioSourceNode();
  }
}

class MediaElementAudioSourceNode {
  connect() {}
}
/* eslint-enable */

// todo 完善这个组件的单元测试
describe('Player component', () => {
  beforeAll(() => {
    if (typeof global.AudioContext !== 'function') {
      global.AudioContext = AudioContext;
    }

    if (typeof global.MediaElementAudioSourceNode !== 'function') {
      global.MediaElementAudioSourceNode = MediaElementAudioSourceNode;
    }
  });

  afterAll(() => {
    if (global.AudioContext === AudioContext) {
      delete global.AudioContext;
    }

    if (global.MediaElementAudioSourceNode === MediaElementAudioSourceNode) {
      delete global.MediaElementAudioSourceNode;
    }
  });

  it('should match exact snapshot', () => {
    const onMusicEnd = spy();
    const player = <Player isPlaying={false} onMusicEnd={onMusicEnd} />;
    const tree = renderer.create(player).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
