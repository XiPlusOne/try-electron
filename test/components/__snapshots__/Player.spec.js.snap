// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Player component should match exact snapshot 1`] = `
<audio
  onEnded={[Function]}
  preload="auto"
  src="D:\\\\Projects\\\\try-electron\\\\resources\\\\bg.flac"
>
  <track
    default={true}
    kind="captions"
    label="Hello World!"
  />
</audio>
`;
